File: Simulado 03/URI/05 - Cops and Robbers.cpp
#include <iostream>
using namespace std;

bool tryPath(int gameMap[5][5], int row, int column, bool tried[5][5]){
  if (gameMap[row][column] == 1)
    return false;
  if ((row == 4) && (column == 4))
    return true;

  tried[row][column] = true;
  bool found = false;


  if ((row < 4) && (!tried[row + 1][column]) && (gameMap[row + 1][column] == 0))
      found |= tryPath(gameMap, row + 1, column, tried);
  if ((!found) && (column < 4) && (!tried[row][column + 1]) && (gameMap[row][column + 1] == 0))
      found |= tryPath(gameMap, row, column + 1, tried);
  if ((!found) && (row >= 1) && (!tried[row - 1][column]) && (gameMap[row - 1][column] == 0))
      found |= tryPath(gameMap, row - 1, column, tried);
  if ((!found) && (column >= 1) && (!tried[row][column - 1]) && (gameMap[row][column - 1] == 0))
      found |= tryPath(gameMap, row, column - 1, tried);
  
  tried[row][column] = false;
  return found;
}

bool tryPath(int gameMap[5][5]){
  bool tried[5][5];
  for (int j = 0; j < 5; j++)
    for (int k = 0; k < 5; k++)
      tried[j][k] = false;
  return tryPath(gameMap, 0, 0, tried);
}

int n;
int gameMap[5][5];

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  cin >> n;

  for (int i = 0; i < n; i++)
  {
    for (int j = 0; j < 5; j++)
      for (int k = 0; k < 5; k++)
        cin >> gameMap[j][k];

    if (tryPath(gameMap))
      cout << "COPS" << endl;
    else
      cout << "ROBBERS" << endl;
  }
  return 0;
}

File: Simulado 03/URI/05 - Cops and Robbers.py
def tryPath(gameMap, row = 0, column = 0, tried = None):
    if gameMap[row][column] == 1:
        return False
    if row == 4 and column == 4:
        return True

    if tried == None:
        tried = [[False for _ in range(5)] for _ in range(5)]
    
    tried[row][column] = True
    found = False

    if (row < 4) and (not tried[row + 1][column]) and (gameMap[row + 1][column] == 0):
        found |= tryPath(gameMap, row + 1, column, tried)
    if (not found) and (column < 4) and (not tried[row][column + 1]) and (gameMap[row][column + 1] == 0):
        found |= tryPath(gameMap, row, column + 1, tried)
    if (not found) and (row >= 1) and (not tried[row - 1][column]) and (gameMap[row - 1][column] == 0):
        found |= tryPath(gameMap, row - 1, column, tried)
    if (not found) and (column >= 1) and (not tried[row][column - 1]) and (gameMap[row][column - 1] == 0):
        found |= tryPath(gameMap, row, column - 1, tried)
    
    tried[row][column] = False
    
    return found

def test():
    n = int(input())

    for _ in range(n):
        input()
        gameMap = [[int(x) for x in input().split()] for _ in range(5)]
        print('COPS' if tryPath(gameMap) else 'ROBBERS')

test()